Add interoperable module guide (#1903)
LiskHQ/lisk-docs: docs/modules/ROOT/pages/build-blockchain/interoperable-module/create-ccm.adoc

Now open the file and import the constants, schemas, and types defined above.

Next, define the following properties of the command:

* `name`: Define a method to get the name of the command and set it to the `CROSS_CHAIN_COMMAND_REACT` constant.
The same name will be used for the cross-chain command which will accept the CCM.
* `schema`: Set the command schema to equal `CCReactCommandParamsSchema`.
* `init()`: To initialize the module, we need access to the methods of the interoperability module.
Update the methods to expect the `interoperabilityMethod` as an argument, and assign it to the private property `_interoperabilityMethod` of the `crossChainReact` command.

.react/commands/react_cc_command.ts
[source,typescript]
----
import {
    BaseCommand,
    CommandVerifyContext,
    CommandExecuteContext,
    VerificationResult,
    VerifyStatus,
    codec,
} from 'lisk-sdk';
import { CROSS_CHAIN_COMMAND_REACT } from '../constants';
import { CCReactCommandParamsSchema, CCReactMessageParamsSchema } from '../schemas';
import { CCReactMessageParams, CCReactCommandParams, InteroperabilityMethod } from '../types';

> export class CrossChainReactCommand extends BaseCommand {

Author: i don't see this followed in the default modules, see here for example: https://github.com/LiskHQ/lisk-sdk/blob/development/framework/src/modules/token/commands/transfer_cross_chain.ts

. `data`: An additional property allowing to send a specific string.
Could be used to send custom data or a reaction, like small replies.

The parameters for the `crossChainReact` command therefore include the <<ccm-params,Parameters for the react CCM>>, and two additional parameters:

.Additional parameters for the crossChainReact command
. `receivingChainID`: The chain ID of the chain the CCM is sent to.
. `messageFee`: The fee for sending the CCM across chains.

Additionally, export corresponding schemas for the parameters of the React CCM.

.react/schemas.ts
[source,typescript]
----
// Schema for the parameters of the reactCrossChain CCM
export const CCReactMessageParamsSchema = {
	// The unique identifier of the schema.
	$id: '/lisk/react/ccReactMessageParams',
	type: 'object',
	// The required parameters for the CCM.
	required: ['reactionType', 'helloMessageID', 'data'],
	// A list describing the required parameters for the CCM.
	properties: {
		reactionType: {
			dataType: 'uint32',
			fieldNumber: 1,
		},
		helloMessageID: {
			dataType: 'string',
			fieldNumber: 2,
		},
		data: {
			dataType: 'string',
			fieldNumber: 3,
			minLength: 0,
			maxLength: 64,
		},
	},
};

> // Schema for the parameters of the react reactCrossChain command

Author: ```suggestion
// Schema for the parameters of the crossChainReact command
```

    } = context;

    const ccReactMessageParams: CCReactMessageParams = {
        reactionType: params.reactionType,
        data: params.data,
        helloMessageID: params.helloMessageID,
    };

    await this._interoperabilityMethod.send(
        context.getMethodContext(),
        senderAddress,
        'hello',
        CROSS_CHAIN_COMMAND_REACT,
        params.receivingChainID,
        params.messageFee,
        codec.encode(CCReactMessageParamsSchema, ccReactMessageParams),
        context.header.timestamp,
    );
}
----

== Final updates of the module and app.ts

Go back to the file `react/module.ts` and update it as described in the code comments.

.react/module.ts
[source,typescript]
----
import { BaseInteroperableModule, ModuleMetadata, ModuleInitArgs } from 'lisk-sdk';
import { CrossChainReactCommand } from './commands/react_cc_command';
import { ReactEndpoint } from './endpoint';
import { ReactMethod } from './method';
import { ReactInteroperableMethod } from './cc_method';
// Import the type for the InteroperabilityMethod
import { InteroperabilityMethod } from './types';

export class ReactModule extends BaseInteroperableModule {

>    	public endpoint = new ReactEndpoint(this.stores, this.offchainStores);

Author: Please explain why you added the three dots

    public crossChainMethod = new ReactInteroperableMethod(this.stores, this.events);

    public metadata(): ModuleMetadata {
        return {
            ...this.baseMetadata(),
            endpoints: [],
            commands: this.commands.map(command => ({
                name: command.name,
                params: command.schema,
            })),
            assets: [],
        };
    }

}
----

Now that the interoperable module is initialized, let's create the command which will be used to create the CCM on the sidechain.

== Add required schemas and constants

Create a new file `react/constants.ts` and export a constant `CROSS_CHAIN_COMMAND_NAME_REACT` representing the name of the new command.

.react/constants.ts
[source,typescript]
----
export const CROSS_CHAIN_COMMAND_NAME_REACT = 'reactCrossChain';
----

Now, let's think about what data exactly we want to send in the CCM.

For the simplicity of this example, the `react` CCM includes the following parameters:

Create a new file `react/schemas.ts` and export a corresponding interface for the CCM parameters.

[#ccm-params]
.Parameters for the react CCM
. `reactionType`: A number indicating the type of the reaction.

> In this example, we will only implement type `0`, which is representing a like.

Author: imho the sentence changes meaning here, and makes it more confusing, so I prefer to keep the original sentence.

> In this example, we will only implement type `0`, which is representing a like.

Author: In the sentence right before, it is mentioned that this is a number, which makes it clear that it is not a string, and also we do not use quotes around the number, which we would do in case it is a string, so imho it is clear enough.

. `helloMessageID`: The ID of the hello message to react to.
As there is always only one Hello message per account maximum, we can use the account address of the sender of the Hello message as unique ID for a Hello message.
. `data`: An additional property allowing to send a specific string.

> Could be used to send custom data or reaction, like small replies.

Reviewer: ```suggestion
Could be used to send custom data or a reaction, like small replies.
```

[source,typescript]
----
import {
    MethodContext,
    ImmutableMethodContext,
    CCMsg,
    ChannelData,
    OwnChainAccount,
} from 'lisk-sdk';

export type TokenID = Buffer;

export interface InteroperabilityMethod {
    getOwnChainAccount(methodContext: ImmutableMethodContext): Promise<OwnChainAccount>;
    send(
        methodContext: MethodContext,
        feeAddress: Buffer,
        module: string,
        crossChainCommand: string,
        receivingChainID: Buffer,
        fee: bigint,
        parameters: Buffer,
        timestamp?: number,
    ): Promise<void>;
    error(methodContext: MethodContext, ccm: CCMsg, code: number): Promise<void>;
    terminateChain(methodContext: MethodContext, chainID: Buffer): Promise<void>;
    getChannel(methodContext: MethodContext, chainID: Buffer): Promise<ChannelData>;
    getMessageFeeTokenID(methodContext: ImmutableMethodContext, chainID: Buffer): Promise<Buffer>;
    getMessageFeeTokenIDFromCCM(methodContext: ImmutableMethodContext, ccm: CCMsg): Promise<Buffer>;
}

----

== Initialize the cross-chain command

Now create a new command called `reactCrossChain`:

[source,bash]
----
lisk generate:command react reactCrossChain
----

> Go the file `react/commands/react_command.ts` and import the constants, schemas and types defined above.

Reviewer: ```suggestion
Go to the file `react/commands/react_command.ts` and import the constants, schemas, and types defined above.
```

Next, define the following properties of the command:

* `name`: Define a method to get the name of the command and set it to the `CROSS_CHAIN_COMMAND_NAME_REACT` constant.
The same name will be used for the cross-chain command which will accept the CCM.
* `schema`: Set the command schema to equal `CCReactCommandParamsSchema`.
* `init()`: To initialize the module, we need access to the methods of the interoperability module.
Update the methods to expect the `interoperabilityMethod` as an argument, and assign it to the private property `_interoperabilityMethod` of the `reactCrossChain` command.

> .react/commands/react_command.ts

Author: It seems this is more a best practice than a requirement? Do we have agreed on certain best practices somewhere? Regarding the naming, this is another question I have: 

Is it required that the command names for commands creating and executing a specific CCM must match? You can also see this question in the code. it would be nice to receive an answer here, and maybe other information about best practices that should be respected in the examples.

    VerifyStatus,
    codec,
} from 'lisk-sdk';
import { CROSS_CHAIN_COMMAND_NAME_REACT } from '../constants';
import {
    CCReactCommandParamsSchema,
    CCReactMessageParams,
    CCReactMessageParamsSchema,
    CCReactCommandParams,
} from '../schemas';
import { InteroperabilityMethod } from '../types';

export class ReactCrossChainCommand extends BaseCommand {
    private _interoperabilityMethod!: InteroperabilityMethod;
    public schema = CCReactCommandParamsSchema;

    public get name(): string {
        return CROSS_CHAIN_COMMAND_NAME_REACT;
    }

    public init(args: { interoperabilityMethod: InteroperabilityMethod }) {
        this._interoperabilityMethod = args.interoperabilityMethod;
    }
}
----

== Command verification

In the `react_command.ts` file, and implement the command verification.

To keep the example simple, we only check if the `receivingChainID` parameter doesn't equal the value of the sending chain.

Extend the `verify()` hook to include more checks for the other parameters as well, as desired.

.react/commands/react_command.ts
[source,typescript]
----
public async verify(context: CommandVerifyContext<Params>): Promise<VerificationResult> {

>         const { params, logger } = context;
> 
>         logger.info('+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++');
>         logger.info(params);
>         logger.info('+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++');
> 
>         try {
>             if (params.receivingChainID.equals(context.chainID)) {
>                 throw new Error('Receiving chain cannot be the sending chain.');
>             }
>         } catch (err) {
>             return {
>                 status: VerifyStatus.FAIL,
>                 error: err as Error,
>             };
>         }
>         return {
>             status: VerifyStatus.OK,
>         };
>     }

Reviewer: ```suggestion
    const { params, logger } = context;

    logger.info('+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++');
    logger.info(params);
    logger.info('+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++');

    try {
        if (params.receivingChainID.equals(context.chainID)) {
            throw new Error('Receiving chain cannot be the sending chain.');
        }
    } catch (err) {
        return {
            status: VerifyStatus.FAIL,
            error: err as Error,
        };
    }
    return {
        status: VerifyStatus.OK,
    };
}
```

----

Once it is verified that the parameters are valid, we can create and send the corresponding CCM.

== Command execution

For this, adjust the `execute()` hook as shown in the snippet below.

Use the `.send()` method of the Interoperability module to send the prepared CCM.

.react/commands/react_command.ts
[source,typescript]
----
public async execute(context: CommandExecuteContext<Params>): Promise<void> {

>         const {
>             params,
>             transaction: { senderAddress },
>         } = context;
> 
>         const reactCCM: CCReactMessageParams = {
>             reactionType: params.reactionType,
>             data: params.data,
>             helloMessageID: params.helloMessageID,
>         };
> 
>         await this._interoperabilityMethod.send(
>             context.getMethodContext(),
>             senderAddress,
>             'hello',
>             CROSS_CHAIN_COMMAND_NAME_REACT,
>             params.receivingChainID,
>             params.messageFee,
>             codec.encode(crossChainReactMessageSchema, reactCCM),
>             context.header.timestamp,
>         );
>     }

Reviewer: ```suggestion
    const {
        params,
        transaction: { senderAddress },
    } = context;

    const reactCCM: CCReactMessageParams = {
        reactionType: params.reactionType,
        data: params.data,
        helloMessageID: params.helloMessageID,
    };

    await this._interoperabilityMethod.send(
        context.getMethodContext(),
        senderAddress,
        'hello',
        CROSS_CHAIN_COMMAND_NAME_REACT,
        params.receivingChainID,
        params.messageFee,
        codec.encode(crossChainReactMessageSchema, reactCCM),
        context.header.timestamp,
    );
}
```

		this.commands[0].init({
			interoperabilityMethod: this._interoperabilityMethod,
		});
	}
}
----

Open the `app.ts` file, and register the module to the application.

Because the `ReactModule` is an *interoperable module*, it is required to call `app.registerInteroperableModule()` additionally.

Last but not least, call the `addDependencies()` method of the `ReactModule` with the methods of the interoperability module as parameter.

IMPORTANT: Please remove the redundant registration of the `ReactModule` in the `modules.ts` file. It was added automatically during the command initialization.

.app.ts
[source,typescript]
----
import { Application, PartialApplicationConfig } from 'lisk-sdk';
import { registerModules } from './modules';
import { registerPlugins } from './plugins';
import { ReactModule } from './modules/react/module';

export const getApplication = (config: PartialApplicationConfig): Application => {
	const { app, method } = Application.defaultApplication(config);
	const reactModule = new ReactModule();
	app.registerModule(reactModule);
	app.registerInteroperableModule(reactModule);

> 	reactModule.addDependencies(method.interoperability);

Author: This should first be fixed in the script that this guide is based on.

> Because the `ReactModule` is an *interoperable module*, it is required to call `app.registerInteroperableModule()` additionally.

Reviewer: ```suggestion
As the `ReactModule` is an *interoperable module*, it is required to call `app.registerInteroperableModule()` additionally.
```

> Last but not least, call the `addDependencies()` method of the `ReactModule` with the methods of the interoperability module as parameter.

Reviewer: ```suggestion
Last but not least, call the `addDependencies()` method of the `ReactModule` with the methods of the interoperability module as a parameter.
```

    registerModules(app);
    registerPlugins(app);

    return app;
};
----

> When a user now posts a `reactCrossChain` transaction on a sidechain that registered the React module, a corresponding CCM is sent to the mainchain by a xref:{url_run_relayer}[relayer node], where it will be forwarded to the designated receiving sidechain.

Reviewer: ```suggestion
When a user posts a `reactCrossChain` transaction on a sidechain that registered the React module, a corresponding CCM is sent to the mainchain by a xref:{url_run_relayer}[relayer node], where it will be forwarded to the designated receiving sidechain.
```
